src/typecheck/thf.ml

```diff
@@ -1315,7 +1331,20 @@ module Make
       end
 
   and parse_pattern_var ty env ast s =
-    let v = mk_term_var env (Id.name s) ty in
+    let check_pattern_var_id (name: Dolmen_std.Name.t) =
+      match name with
+      | Simple s
+      | Indexed { basename = s; _ }
+      | Qualified { basename = s; _ } ->
+        let c = String.get s 0 in
+        (* Check if the name starts with an uppercase character. *)
+        if int_of_char c >= 65 && int_of_char c <= 90
```

Comment:
This whole check is very very wrong. In the typechecker, you cannot suppose that "constructor=first letter capitalised": it's a property of a language, and the typechecker has to correctly handle different languages.
If I understand correctly, the problem is that without this piece of code, using a not-defined constructor with no arguments in a pattern matching would lead to it being understood as a variable, right ? If so, the answer is more likely to introduce a namespace for constructors, and ensure that symbols with that namespace are not turned into variables. (the other way around is not reliable, as most language cannot statically identifiy what is or is not a variable).

Reply:
Ok, my bad. I assumed that it was the case (that "constructor=first letter capitalized"). But since it isn't I will just remove it. I don't think it's necessary to add a special namespace for constructors since the call to ```find_bound``` guarantees that if the symbol represents a constructor, it will not be considered as a variable.

Reply:
Just as an example, in tptp, variables must all start with an uppercase letter, ^^

Reply:
Oups...

Reply:
I guess you'll see but if you just remove this check, i'm afraid your example with the unbound `C` constructor will now pass probably with a redundant error on the last case, because `C` will be turned into a variable (unless something prevents that).

Reply:
The behaviour is the same in Alt-Ergo, ```C``` is considered as a variable and the following cases are considered as redundant. Here's another [example](https://github.com/OCamlPro/alt-ergo/blob/next/non-regression/invalid/adts/useless-patterns-vars.ae).
```
[Warning] File "non-regression/invalid/adts/useless-patterns-vars.ae", line 5, characters 1-58:Typing Error: Pattern-matching contains unreachable cases. These cases are removed: sB | sC
File "non-regression/invalid/adts/useless-patterns-vars.ae", line 4, characters 1-77: I don't know (0.0025) (4 steps) (goal g_invalid)
```
```D``` is considered as a variable. 
But AE just removes the useless cases and responds, not sure how to turn the Error into a warning in Dolmen, I am also not sure if it's better to have it as an error or as a warning.

Reply:
Well, if alt-ergo allows variable to start with an uppercase letter, then indeed, there's not much we can do here.